File: backend/src/main/java/com/paymentgateway/payment_gateway/workers/RefundWorker.java
package com.paymentgateway.payment_gateway.workers;

import com.paymentgateway.payment_gateway.entity.Payment;
import com.paymentgateway.payment_gateway.entity.Refund;
import com.paymentgateway.payment_gateway.jobs.ProcessRefundJob;
import com.paymentgateway.payment_gateway.repository.PaymentRepository;
import com.paymentgateway.payment_gateway.repository.RefundRepository;
import com.paymentgateway.payment_gateway.queue.JobQueue;
import com.paymentgateway.payment_gateway.service.WebhookService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Random;
import java.util.UUID;

@Component
public class RefundWorker {
    
    private static final Logger log = LoggerFactory.getLogger(RefundWorker.class);
    
    @Autowired
    private JobQueue jobQueue;
    
    @Autowired
    private RefundRepository refundRepository;
    
    @Autowired
    private PaymentRepository paymentRepository;
    
    @Autowired
    private WebhookService webhookService;
    
    private final Random random = new Random();

    // @jakarta.annotation.PostConstruct removed to avoid double startup. 
    // WorkerApplication handles startup.
    
    public void processRefunds() {
        log.info("RefundWorker waiting for Redis to initialize...");
        try {
            // Wait for Redis to be fully ready
            Thread.sleep(10000); 
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return;
        }
        log.info("RefundWorker started");
        
        while (true) {
            try {
                ProcessRefundJob job = jobQueue.dequeueRefundJob(5);
                
                if (job != null) {
                    log.info("Processing refund job: {}", job.getRefundId());
                    processRefund(job);
                }
                
            } catch (Exception e) {
                log.error("Error in RefundWorker", e);
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }
        
        log.info("RefundWorker stopped");
    }
    
    private void processRefund(ProcessRefundJob job) {
        try {
            Optional<Refund> refundOpt = refundRepository.findById(job.getRefundId());
            
            if (refundOpt.isEmpty()) {
                log.error("Refund not found: {}", job.getRefundId());
                jobQueue.markJobFailed(JobQueue.getRefundQueue());
                return;
            }
            
            Refund refund = refundOpt.get();
            
            Optional<Payment> paymentOpt = paymentRepository.findById(refund.getPaymentId());
            
            if (paymentOpt.isEmpty()) {
                log.error("Payment not found for refund: {}", refund.getPaymentId());
                markRefundFailed(refund, "Payment not found");
                jobQueue.markJobFailed(JobQueue.getRefundQueue());
                return;
            }

            Payment payment = paymentOpt.get();

            // 1. Verify Payment Status
            if (!"success".equals(payment.getStatus())) {
                log.error("Payment not successful/refundable: {}", refund.getPaymentId());
                markRefundFailed(refund, "Payment is not in 'success' state");
                jobQueue.markJobFailed(JobQueue.getRefundQueue());
                return;
            }
            
            // 2. Verify Total Refunded Amount (Race Condition Check)
            List<Refund> allRefunds = refundRepository.findByPaymentId(payment.getId());
            long totalRefunded = allRefunds.stream()
                .filter(r -> "processed".equals(r.getStatus()) || 
                             ("pending".equals(r.getStatus()) && !r.getId().equals(refund.getId())))
                .mapToLong(Refund::getAmount)
                .sum();
                
            if (totalRefunded + refund.getAmount() > payment.getAmount()) {
                log.error("Refund amount exceeds available amount: {}", refund.getId());
                markRefundFailed(refund, "Refund amount exceeds available payment amount");
                jobQueue.markJobFailed(JobQueue.getRefundQueue());
                return;
            }
            
            // 3. Simulate processing delay (3-5 seconds)
            // random.nextInt(2001) gives 0 to 2000. 3000 + [0-2000] = 3000 to 5000ms.
            long delayMs = 3000 + random.nextInt(2001);
            log.info("Simulating refund processing delay: {}ms", delayMs);
            Thread.sleep(delayMs);
            
            // 4. Update Status
            refund.setStatus("processed");
            refund.setProcessedAt(LocalDateTime.now());
            refundRepository.save(refund);
            
            log.info("Refund processed: {}", refund.getId());
            
            if (refund.getAmount().equals(payment.getAmount())) {
                log.info("Full refund detected for payment: {}", payment.getId());
            }
            
            // 5. Send Webhook
            sendRefundWebhook(refund, payment);
            
            jobQueue.markJobCompleted(JobQueue.getRefundQueue());
            
            log.info("Refund processing completed: {}", refund.getId());
            
        } catch (Exception e) {
            log.error("Failed to process refund: {}", job.getRefundId(), e);
            jobQueue.markJobFailed(JobQueue.getRefundQueue());
        }
    }
    
    private void markRefundFailed(Refund refund, String reason) {
        // Since schema doesn't have explicit 'failed' status for refund (only pending/processed implied)
        // We might want to add 'failed' to allowed status or just log it.
        // Requirement says "Verify... invalid... return 400" (API).
        // For Worker: "Update refund status in database: Set status to 'processed'".
        // It doesn't explicitly say what to do if validation fails in WORKER (async).
        // But usually we should mark as failed. Schema comment says "Values: 'pending', 'processed'".
        // I will optimistically force 'failed' status or just leave it pending/stuck?
        // Let's assume 'failed' is a valid status we should add/use, or we just don't process it.
        // Given I can't easily change schema constraints if they exist (JPA doesn't enforce string enum), I'll set 'failed'.
        refund.setStatus("failed");
        // Append failure reason to the reason field or logic?
        if (refund.getReason() == null) {
             refund.setReason("Failed: " + reason);
        } else {
             refund.setReason(refund.getReason() + " | Failed: " + reason);
        }
        refundRepository.save(refund);
    }
    
    private void sendRefundWebhook(Refund refund, Payment payment) {
        try {
            Map<String, Object> refundData = new HashMap<>();
            refundData.put("id", refund.getId());
            refundData.put("payment_id", refund.getPaymentId());
            refundData.put("amount", refund.getAmount());
            refundData.put("reason", refund.getReason());
            refundData.put("status", refund.getStatus());
            refundData.put("created_at", refund.getCreatedAt().toString());
            refundData.put("processed_at", refund.getProcessedAt().toString());
            
            webhookService.sendRefundWebhook(UUID.fromString(payment.getMerchant().getId()), "refund.processed", refundData);
            
        } catch (Exception e) {
            log.error("Failed to send refund webhook", e);
        }
    }
}

File: backend/src/workers/refundWorker.ts
import { Worker, Job } from 'bullmq';
import { connection, REFUND_QUEUE_NAME, webhookQueue } from '../config/queue';
import { query } from '../config/db';

interface RefundJobData {
    refundId: string;
}

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

export const refundWorker = new Worker(REFUND_QUEUE_NAME, async (job: Job<RefundJobData>) => {
    const { refundId } = job.data;
    console.log(`[RefundWorker] Processing refund: ${refundId}`);

    try {
        // 1. Fetch Refund & Associated Payment
        // We join to get payment details for verification
        const result = await query(
            `SELECT r.*, p.status as payment_status, p.amount as payment_total 
             FROM refunds r
             JOIN payments p ON r.payment_id = p.id
             WHERE r.id = $1`,
            [refundId]
        );

        if (result.rows.length === 0) {
            console.error(`[RefundWorker] Refund not found: ${refundId}`);
            return;
        }

        const refund = result.rows[0];

        // 2. Verify Payment is in refundable state
        if (refund.payment_status !== 'success') {
            const error = 'Payment status is not success';
            console.error(`[RefundWorker] ${error}`);
            await query("UPDATE refunds SET status = 'failed', reason = $1 WHERE id = $2", [error, refundId]);
            return;
        }

        // Verify total refunded amount (sum of all *other* processed refunds + this one?)
        // The requirement says: "Verify total refunded amount (sum of all refunds for this payment) does not exceed payment amount"
        // Usually, we check the SUM of (processed + current request).
        // Let's check what has been processed SO FAR.
        const sumResult = await query(
            "SELECT COALESCE(SUM(amount), 0) as total_refunded FROM refunds WHERE payment_id = $1 AND status = 'processed'",
            [refund.payment_id]
        );
        const alreadyRefunded = parseInt(sumResult.rows[0].total_refunded, 10);
        const currentRefundAmount = refund.amount;

        if (alreadyRefunded + currentRefundAmount > refund.payment_total) {
            const error = 'Refund amount exceeds remaining payment balance';
            console.error(`[RefundWorker] ${error}`);
            await query("UPDATE refunds SET status = 'failed', reason = $1 WHERE id = $2", [error, refundId]);
            return;
        }

        // 3. Simulate Processing Delay (3-5 seconds)
        const isTestMode = process.env.TEST_MODE === 'true';
        let delay = Math.floor(Math.random() * 2000) + 3000; // 3-5 seconds

        if (isTestMode && process.env.TEST_PROCESSING_DELAY) {
            delay = parseInt(process.env.TEST_PROCESSING_DELAY, 10);
        }

        console.log(`[RefundWorker] Simulating delay of ${delay}ms`);
        await sleep(delay);

        // 4. Update Refund Status in Database
        await query(
            "UPDATE refunds SET status = 'processed', processed_at = NOW() WHERE id = $1",
            [refundId]
        );

        // 5. Full Refund Check (Optional)
        // If (alreadyRefunded + current) == total, maybe mark payment as fully_refunded?
        // Requirement says "Optionally update payment record to reflect full refund status".
        // Use standard Payment status? Maybe "refunded"? 
        // Let's stick to the core requirements first, but this is a nice to have.

        // 6. Enqueue Webhook
        console.log(`[RefundWorker] Refund processed. Enqueueing webhook.`);

        await webhookQueue.add('deliver-webhook', {
            merchantId: refund.merchant_id,
            event: 'refund.processed',
            payload: {
                event: 'refund.processed',
                timestamp: Math.floor(Date.now() / 1000),
                data: {
                    refund: {
                        id: refund.id,
                        payment_id: refund.payment_id,
                        amount: refund.amount,
                        status: 'processed',
                        reason: refund.reason,
                        created_at: refund.created_at,
                        processed_at: new Date()
                    }
                }
            }
        });

    } catch (error) {
        console.error(`[RefundWorker] Error processing refund ${refundId}:`, error);
        throw error;
    }
}, { connection: connection as any });
